### DISCUSSION TITLE
Input & Output file format

### DISCUSSION THREAD

**Discussion by sidml** [ZigRazor/CXXGraph#170] [Ideas]
I was thinking that we should ensure compatibility with existing popular formats for saving and loading graphs. Following is taken from [here](https://www.boost.org/doc/libs/1_77_0/libs/graph/doc/)

>  The BGL currently provides two graph classes and an edge list adaptor:
    adjacency_list
    adjacency_matrix
    edge_list

> The adjacency_list class is the general purpose “swiss army knife” of graph classes. It is highly parameterized so that it can be optimized for different situations: the graph is directed or undirected, allow or disallow parallel edges, efficient access to just the out-edges or also to the in-edges, fast vertex insertion and removal at the cost of extra space overhead, etc.

>The adjacency_matrix class stores edges in a |V| x |V| matrix (where |V| is the number of vertices). The elements of this matrix represent edges in the graph. Adjacency matrix representations are especially suitable for very dense graphs, i.e., those where the number of edges approaches |V|2.

> The edge_list class is an adaptor that takes any kind of edge iterator and implements an Edge List Graph. 

Let me know what you guys think.
@ZigRazor @AlfredCP @pavan-pan

**Comment by sidml** [comment]
Few example datasets in different formats are [here](https://graphia.app/example-data.html).

**Comment by ZigRazor** [comment]
I think is a beatiful idea. 

I thought long ago about this, but solo is difficult to collect and implement all format.
I think should be a new functionalities.
Maybe we can start from one format and then add the others as we go on.

**Comment by AlfredCP** [comment]
I am not completely familiar with graphs theory but more on coding.

As far as I understand we want something like:


 xml || csv || txt || ....  input file  -------> Graph or Graph_TS  ------>  algorithms ------> Result  ------>   xml || csv || txt || ....  output file


Right?
I mean, have a Parser class that derives in different children for each of the Input formats.
Then common code for algorithm processing
Finally a Formatter that derives for each output format
Is this the idea??

If so, it sound good, yes. I think it needs a proper design since will be a substantial refactoring in case we need to modify the Graph class (maybe not needed)

**Comment by AlfredCP** [comment]
Let's say, make algorithms and core logic totally independent from input and output formats, right?

**Reply by ZigRazor** [reply]
> Let's say, make algorithms and core logic totally independent from input and output formats, right?

Yes, I think we need to completely separate core from input and output ( as it is now )
Maybe, a class for parsing could be useful, for varius format of input, and a class formatter can be useful for output, separated from the Graph class.

**Comment by ZigRazor** [comment]
[Here](https://snap.stanford.edu/data/) there are some good graph dataset. I think we need to start from these dataset and provide a reader for each input format and a formatter for each format. 

These dataset are good also to test the algorithm, is a good idea ( after the implementation of the import/export ) to test the algorithm with these dataset

**Comment by AlfredCP** [comment]
I think this is a really good idea. The fact of making the library easy to use for the final user would boost the usage (it is also a lot of work, but just matter of time).

I like it!

**Comment by ZigRazor** [comment]
if you agree I open an Issue to start the work.
I think the first step is the definition of the interface

**Reply by sidml** [reply]
Please open the issue. Then we can start discussing the concrete details about the implementation.

**Reply by ZigRazor** [reply]
Opened
